### PR TITLE
Add group-based granular authorization for individual routes.

### PR DESCRIPTION
**Pull request by spladug**
:eyeglasses: @JordanMilne

### CODE REVIEW

File: underpants.sample.groups.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  "certs" : [
+    {
+      "crt" : "/path/to/crt.pem",
+      "key" : "/path/to/key.pem"
+    }
+  ],
+  "routes" : [
+    {
+      "from"           : "public.company.com",
+      "to"             : "http://localhost:8080",
```
Comment (JordanMilne) [inline underpants.sample.groups.json:24]: Is there any way someone could cause underpants to make a request to `http://localhost:8080@whatever.com/`?

File: underpants.go [hunk 2]
```diff
@@ -257,6 +277,31 @@ func serveHttpProxy(d *disp, w http.ResponseWriter, r *http.Request) {
     return
   }
 
+  if d.config.HasGroups() {
+    authorized := false
+
+AuthCheck:
```
Comment (JordanMilne) [inline underpants.go:283]: This tripped me up and made me think the `break` would jump back up here. Can you move this under the `for`s or is it a style thing?

File: underpants.sample.groups.json (hunk 1 above)
Comment (spladug) [inline underpants.sample.groups.json:24]: Tracing through, let me know if any of this sounds wrong.
- `setup` parses the `To` URL and extracts the host and scheme, ignoring all else. (https://github.com/kellegous/underpants/blob/master/underpants.go#L418)
- the host and scheme are passed onto the `disp` object which handles requests for that route (https://github.com/kellegous/underpants/blob/master/underpants.go#L425)
- `serveHttpProxy` builds a new request to the by calling `urlFor` (https://github.com/kellegous/underpants/blob/master/underpants.go#L260)
- `urlFor` takes the requested URL and changes its host and scheme to those configured for the backend (https://github.com/kellegous/underpants/blob/master/underpants.go#L244-L247)

_so_, it's possible only if go's urlparser reserializes poorly.  whatcha think?

File: underpants.go (hunk 2 above)
Comment (spladug) [inline underpants.go:283]: My, very limited, understanding of labels in Go is that they need to come before the statement you're labelling.
- http://golang.org/ref/spec#Break_statements
- http://golang.org/ref/spec#Label

File: underpants.sample.groups.json (hunk 1 above)
Comment (JordanMilne) [inline underpants.sample.groups.json:24]: Sounds like an excuse for me to try breaking a new URL parser / serializer

Comment (spladug) [inline underpants.sample.groups.json:24]: Excellent.

FWIW, this URL stuff only happens after authentication, so it'd have to be someone with access to a google account already. Somewhat more limited scope, thankfully.

**Pull request closed without merging**